File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/src/main/kotlin/com/runanywhere/sdk/public/events/EventBus.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * Simple pub/sub for SDK events.
 *
 * Mirrors Swift EventBus.swift exactly: subscribes to the canonical native
 * SDKEvent stream via rac_sdk_event_subscribe, decodes the proto bytes, and
 * fans them into a Kotlin MutableSharedFlow so consumers see lifecycle,
 * model, error, and other events emitted by C++ commons.
 */

package com.runanywhere.sdk.public.events

import ai.runanywhere.proto.v1.EventCategory.EVENT_CATEGORY_ERROR
import ai.runanywhere.proto.v1.EventCategory.EVENT_CATEGORY_LLM
import ai.runanywhere.proto.v1.EventCategory.EVENT_CATEGORY_MODEL
import ai.runanywhere.proto.v1.EventCategory.EVENT_CATEGORY_RAG
import ai.runanywhere.proto.v1.EventCategory.EVENT_CATEGORY_SDK
import ai.runanywhere.proto.v1.EventCategory.EVENT_CATEGORY_STT
import ai.runanywhere.proto.v1.EventCategory.EVENT_CATEGORY_TTS
import com.runanywhere.sdk.infrastructure.logging.SDKLogger
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.mapNotNull
import kotlinx.coroutines.launch

/**
 * Central event bus for SDK-wide event distribution.
 *
 * Subscribe to events by category or to all events:
 *
 * ```kotlin
 * // Subscribe to all events
 * EventBus.events.collect { event ->
 *     println(event.type)
 * }
 *
 * // Subscribe to specific category
 * EventBus.events(EventCategory.LLM).collect { event ->
 *     println(event.type)
 * }
 * ```
 *
 * Mirrors Swift EventBus exactly. Native SDK events emitted by C++ commons
 * (model lifecycle, errors, init, device, etc.) flow into this bus through
 * the [start] / [stop] lifecycle hooks called by the platform bridge during
 * SDK init/shutdown.
 */
object EventBus {
    // MARK: - Publishers

    private val logger = SDKLogger.shared

    private val _events =
        MutableSharedFlow<SDKEvent>(
            replay = 0,
            extraBufferCapacity = 64,
        )

    /** All events flow */
    val events: Flow<SDKEvent> = _events.asSharedFlow()

    // MARK: - Native subscription lifecycle

    /**
     * Start the native SDK event subscription. Idempotent: calling twice
     * is a no-op. Invoked by the platform bridge after C++ commons is
     * initialized so native events are delivered into [events].
     */
    fun start() {
        startNativeSubscription()
    }

    /**
     * Stop the native SDK event subscription. Idempotent: calling twice
     * is a no-op. Invoked by the platform bridge during shutdown so the
     * native side releases the subscription slot.
     */
    fun stop() {
        stopNativeSubscription()
    }

    /**
     * Internal entry point used by the platform bridge to push a native
     * SDKEvent into the bus. Non-suspending; safe to call from any
     * thread (including the JNI callback thread).
     */
    internal fun emitFromNative(event: SDKEvent) {
        _events.tryEmit(event)
    }

    // MARK: - Publishing

    /**
     * Publish an event to all subscribers.
     *
     * Mirrors Swift: tries the native publish path first so the canonical
     * stream sees the event (and re-delivers it back through the
     * subscription); falls back to a direct local emit if native is not
     * available (e.g. native lib not loaded).
     */
    fun publish(event: SDKEvent) {
        logger.debug("Publishing event: ${event.id} (category: ${event.category})")
        if (!publishToNative(event)) {
            _events.tryEmit(event)
        }
    }

    // MARK: - Filtered Subscriptions

    /**
     * Get events for a specific category.
     */
    fun events(category: EventCategory): Flow<SDKEvent> {
        return events.filter { it.category == category }
    }

    /**
     * Subscribe to all events with a closure. The returned [Job] is the
     * subscription token; cancel it to stop receiving events.
     */
    fun on(
        scope: CoroutineScope,
        handler: suspend (SDKEvent) -> Unit,
    ): Job =
        scope.launch {
            events.collect { event -> handler(event) }
        }

    /**
     * Subscribe to events of a specific category with a closure. The returned
     * [Job] is the subscription token; cancel it to stop receiving events.
     */
    fun on(
        scope: CoroutineScope,
        category: EventCategory,
        handler: suspend (SDKEvent) -> Unit,
    ): Job =
        scope.launch {
            events(category).collect { event -> handler(event) }
        }

    /**
     * Extract a specific payload type from the proto envelope stream for
     * internal convenience streams.
     */
    private fun <T : Any> eventsOfPayload(selector: (SDKEvent) -> T?): Flow<T> = events.mapNotNull(selector)

    /** Stream of [VoiceEvent] payloads (voice-agent pipeline events). */
    val voiceEventPayloads: Flow<VoiceEvent>
        get() = eventsOfPayload { it.voice_pipeline }

    /** Stream of [DownloadEvent] payloads (model download progress / lifecycle). */
    val downloadEventPayloads: Flow<DownloadEvent>
        get() = eventsOfPayload { it.download }

    /** Stream of [ComponentLifecycleEvent] payloads. */
    val componentLifecycleEventPayloads: Flow<ComponentLifecycleEvent>
        get() = eventsOfPayload { it.component_lifecycle }

    /** Stream of [ModelRegistryEvent] payloads. */
    val modelRegistryEventPayloads: Flow<ModelRegistryEvent>
        get() = eventsOfPayload { it.model_registry }

    // MARK: - Convenience Methods

    /**
     * Get LLM events.
     */
    val llmEvents: Flow<SDKEvent>
        get() = events(EVENT_CATEGORY_LLM)

    /**
     * Get STT events.
     */
    val sttEvents: Flow<SDKEvent>
        get() = events(EVENT_CATEGORY_STT)

    /**
     * Get TTS events.
     */
    val ttsEvents: Flow<SDKEvent>
        get() = events(EVENT_CATEGORY_TTS)

    /**
     * Get model events.
     */
    val modelEvents: Flow<SDKEvent>
        get() = events(EVENT_CATEGORY_MODEL)

    /**
     * Get error events.
     */
    val errorEvents: Flow<SDKEvent>
        get() = events(EVENT_CATEGORY_ERROR)

    /**
     * Get SDK lifecycle events.
     */
    val sdkEvents: Flow<SDKEvent>
        get() = events(EVENT_CATEGORY_SDK)

    /**
     * Get RAG events.
     */
    val ragEvents: Flow<SDKEvent>
        get() = events(EVENT_CATEGORY_RAG)
}

// Platform-specific native bridge hooks (expect/actual pattern)

File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Public/Events/EventBus.swift
//
//  EventBus.swift
//  RunAnywhere SDK
//
//  Combine publisher over canonical SDK proto events.
//

import Combine
import os

// MARK: - Event Bus

/// Central publisher for SDK-wide `RASDKEvent` distribution.
///
/// Subscribe to events by category or to all events:
/// ```swift
/// // Subscribe to all events
/// EventBus.shared.events
///     .sink { event in print(event.category) }
///
/// // Subscribe to specific category
/// EventBus.shared.events(for: .llm)
///     .sink { event in print(event.properties) }
/// ```
public final class EventBus: @unchecked Sendable {

    // MARK: - Singleton

    public static let shared = EventBus()

    // MARK: - Publishers

    private let subject = PassthroughSubject<RASDKEvent, Never>()

    /// All events publisher
    public var events: AnyPublisher<RASDKEvent, Never> {
        subject.eraseToAnyPublisher()
    }

    private let nativeSubscriptionId = OSAllocatedUnfairLock<UInt64>(initialState: 0)

    // MARK: - Initialization

    private init() {}

    // No `deinit`: `shared` is the only allocation site and it lives for the
    // process lifetime. The native subscription is owned by the explicit
    // `start()` / `stop()` lifecycle below.

    // MARK: - Native subscription lifecycle

    /// Start the native SDK event subscription. Idempotent: calling twice is a
    /// no-op. Invoked by `CppBridge.initialize` after the C++ commons core is
    /// brought up so native lifecycle/model/error events flow into `events`.
    ///
    /// Subscribing before `start()` is safe, but events emitted before the
    /// native subscription is wired are not delivered.
    public func start() {
        nativeSubscriptionId.withLock { id in
            guard id == 0 else { return }
            id = CppBridge.Events.subscribeSDKEvents { [weak self] event in
                self?.subject.send(event)
            }
        }
    }

    /// Stop the native SDK event subscription. Idempotent: calling twice is a
    /// no-op. Invoked during shutdown before the C++ commons core is torn down
    /// so the unsubscribe call still has a working native ABI surface.
    public func stop() {
        nativeSubscriptionId.withLock { id in
            guard id != 0 else { return }
            CppBridge.Events.unsubscribeSDKEvents(id)
            id = 0
        }
    }

    // MARK: - Publishing

    /// Publish an event to all subscribers
    public func publish(_ event: RASDKEvent) {
        if !CppBridge.Events.publishSDKEvent(event) {
            subject.send(event)
        }
    }

    // MARK: - Filtered Subscriptions

    /// Get events for a specific category
    public func events(for category: RAEventCategory) -> AnyPublisher<RASDKEvent, Never> {
        subject
            .filter { $0.category == category }
            .eraseToAnyPublisher()
    }

    /// Extract a specific payload type from the proto envelope stream for
    /// internal convenience streams.
    private func eventsOfPayload<Payload>(
        _ selector: @escaping (RASDKEvent) -> Payload?
    ) -> AnyPublisher<Payload, Never> {
        subject
            .compactMap(selector)
            .eraseToAnyPublisher()
    }

    /// Stream of `RAVoiceEvent` payloads (voice-agent pipeline events).
    public var voiceEventPayloads: AnyPublisher<RAVoiceEvent, Never> {
        eventsOfPayload { envelope in
            guard case .voicePipeline(let payload)? = envelope.event else { return nil }
            return payload
        }
    }

    /// Stream of `RADownloadEvent` payloads (model download progress / lifecycle).
    public var downloadEventPayloads: AnyPublisher<RADownloadEvent, Never> {
        eventsOfPayload { envelope in
            guard case .download(let payload)? = envelope.event else { return nil }
            return payload
        }
    }

    /// Stream of `RAComponentLifecycleEvent` payloads.
    public var componentLifecycleEventPayloads: AnyPublisher<RAComponentLifecycleEvent, Never> {
        eventsOfPayload { envelope in
            guard case .componentLifecycle(let payload)? = envelope.event else { return nil }
            return payload
        }
    }

    /// Stream of `RAModelRegistryEvent` payloads.
    public var modelRegistryEventPayloads: AnyPublisher<RAModelRegistryEvent, Never> {
        eventsOfPayload { envelope in
            guard case .modelRegistry(let payload)? = envelope.event else { return nil }
            return payload
        }
    }

    // MARK: - Convenience Methods

    /// Get LLM events.
    public var llmEvents: AnyPublisher<RASDKEvent, Never> {
        events(for: .llm)
    }

    /// Get STT events.
    public var sttEvents: AnyPublisher<RASDKEvent, Never> {
        events(for: .stt)
    }

    /// Get TTS events.
    public var ttsEvents: AnyPublisher<RASDKEvent, Never> {
        events(for: .tts)
    }

    /// Get model events.
    public var modelEvents: AnyPublisher<RASDKEvent, Never> {
        events(for: .model)
    }

    /// Get error events.
    public var errorEvents: AnyPublisher<RASDKEvent, Never> {
        events(for: .error)
    }

    /// Get SDK lifecycle events.
    public var sdkEvents: AnyPublisher<RASDKEvent, Never> {
        events(for: .sdk)
    }

    /// Get RAG events.
    public var ragEvents: AnyPublisher<RASDKEvent, Never> {
        events(for: .rag)
    }

    // MARK: - Closure Subscriptions

    /// Subscribe to events with a closure
    public func on(_ handler: @escaping (RASDKEvent) -> Void) -> AnyCancellable {
        subject.sink { event in
            handler(event)
        }
    }

    /// Subscribe to events of a specific category
    public func on(
        _ category: RAEventCategory,
        handler: @escaping (RASDKEvent) -> Void
    ) -> AnyCancellable {
        events(for: category).sink { event in
            handler(event)
        }
    }
}
